mlflow/utils/async_logging/run_artifact.py

```diff
@@ -1,32 +1,59 @@
+import posixpath
+import tempfile
 import threading
-from typing import TYPE_CHECKING, Union
-
-if TYPE_CHECKING:
-    import PIL
+from typing import Optional
 
 
 class RunArtifact:
     def __init__(
         self,
         filename: str,
         artifact_path: str,
-        artifact: Union["PIL.Image.Image"],
-        completion_event: threading.Event,
+        artifact: Optional["PIL.Image.Image"] = None,
+        completion_event: Optional[threading.Event] = None,
+        local_dir: Optional[str] = None,
     ) -> None:
         """Initializes an instance of `RunArtifacts`.
 
         Args:
-            filename: Filename of the artifact to be logged
+            filename: Real filename of the artifact to be logged
             artifact_path: Directory within the run's artifact directory in which to log the
                 artifact.
-            artifact: The artifact to be logged.
+            artifact (Deprecated): A PIL image to be logged.
             completion_event: A threading.Event object.
+            local_dir: Local directory in which the artifact is or will be stored. If not provided,
+                a temporary directory will be created and used.
         """
-        self.filename = filename
+        self._tmpdir = None
         self.artifact_path = artifact_path
-        self.artifact = artifact
-        self.completion_event = completion_event
+        self.filename = filename
+        self.image = artifact
+        if local_dir is None:
+            self._tmpdir = tempfile.TemporaryDirectory()
+            local_dir = self._tmpdir.name
+
+        self.local_filepath = posixpath.join(local_dir, filename)
+        self.completion_event = threading.Event()
         self._exception = None
+        self.completion_event = threading.Event() if completion_event is None else completion_event
+        self._handle_image()
+
+    def _handle_image(self):
+        if self.image is not None:
+            from PIL import Image
+            if isinstance(self.image, Image.Image):
+                self.image.save(self.local_filepath)
```

Comment:
This logic should not exist in RunArtifact, but should exist in log_image

Reply:
Could we revert the changes modified in RunArtifact and update _async_artifact_logging_function to handle different artifacts?